src/GcodeLayerThreader.h

```diff
@@ -0,0 +1,213 @@
+/** Copyright (C) 2017 Ultimaker - Released under terms of the AGPLv3 License */
+#ifndef GCODE_LAYER_THREADER_H
+#define GCODE_LAYER_THREADER_H
+
+#include <queue> // priority_queue
+#include <functional> // function
+#include <thread> // sleep
+#include <chrono> // milliseconds
+
+#include "utils/logoutput.h"
+#include "utils/optional.h"
+#include "utils/Lock.h"
+
+namespace cura
+{
+
+/*!
+ * Producer Consumer construct for when:
+ * - production can occur in parallel
+ * - consumption must be ordered and not multithreaded
+ * 
+ * A layer_nr index is passed to the item producer in order to produce the different items.
+ * 
+ * Each thread does production and consumption, giving priority to consumption if some is available.
+ * 
+ * If there is only one thread, it consumes every time it has produced one item.
+ * 
+ * \warning This class is only adequate when the expected production time of an item is more than (n_threads - 1) times as much as the expected consumption time of an item
+ */
+template <typename T>
+class GcodeLayerThreader
```

Comment:
I did that on purpose. The functionality is very specific to the problem of multithreading gcode writing.

This class works for every pipeline where the production is ordered and each item produced can be identified by an integer in a consecutive list and the production can be performed in parallel, while the consumption cannot be performed in parallel. This class also minimizes the amount of items which are being / have been produced but haven't been consumed yet.
All this is so specific to our problem at hand that I found it weird to give the class a general name.